### PR TITLE
Split connection management away from exporter

### PR DESCRIPTION
**Pull request by krnowak**
This commits adds an ConnectionManager interface, which the exporter
will use to connect to the collector and send both metrics and traces
to it. That way, the Exporter type is free from dealing with any
connection/protocol details, as this business is taken over by the
implementations of the ConnectionManager interface.

The gRPC code from the exporter is moved into the implementation of
ConnectionManager. Currently it only maintains a single connection,
just as the Exporter used to do.

With the split, most of the Exporter options became actually gRPC
connection manager's options. Currently the only option that remained
to be Exporter's is about setting the export kind selector.

### CODE REVIEW

File: exporters/otlp/connectionmanager.go [hunk 1]
```diff
@@ -0,0 +1,51 @@
...
+	// ExportMetrics should transform the passed metrics to the
+	// wire format and send it to the collector. May be called
+	// concurrently with ExportTraces, so the manager needs to
+	// take this into account by doing proper locking.
+	ExportMetrics(parent context.Context, cps metricsdk.CheckpointSet, selector metricsdk.ExportKindSelector) error
+	// ExportTraces should transform the passed traces to the wire
+	// format and send it to the collector. May be called
+	// concurrently with ExportMetrics, so the manager needs to
+	// take this into account by doing proper locking.
+	ExportTraces(ctx context.Context, sds []*tracesdk.SpanData) error
```
Comment (MrAlias) [inline exporters/otlp/connectionmanager.go:50]: Should these instead take the proto format of the data so this interface is entirely responsible for the connection and transport of data instead of also the transformation, which the Exporter itself will be responsible for?

File: exporters/otlp/connectionmanager.go [hunk 2]
```diff
@@ -0,0 +1,51 @@
...
+	// concurrently with ExportMetrics or ExportTraces, so proper
+	// locking is required. The function serves as a
+	// synchronization point - after the function returns, the
+	// process of closing connections is assumed to be finished.
+	Stop(ctx context.Context) error
+	// ExportMetrics should transform the passed metrics to the
+	// wire format and send it to the collector. May be called
+	// concurrently with ExportTraces, so the manager needs to
+	// take this into account by doing proper locking.
+	ExportMetrics(parent context.Context, cps metricsdk.CheckpointSet, selector metricsdk.ExportKindSelector) error
```
Comment (MrAlias) [inline exporters/otlp/connectionmanager.go:45]: Given this doesn't return a context, can we unify the naming here to be `ctx`?

```suggestion
	ExportMetrics(ctx context.Context, cps metricsdk.CheckpointSet, selector metricsdk.ExportKindSelector) error
```

**Review by MrAlias — COMMENTED** [review]
I like the overall design of adding the ConnectionManager interface. I'm interested to know what you think about leaving the transform functionality in the exporter though and keeping the new ConnectionManager interface focused on transport.

File: exporters/otlp/connectionmanager.go (hunk 2 above)
Comment (krnowak) [reply exporters/otlp/connectionmanager.go:45]: This must have been a copy pasta, will change it.

File: exporters/otlp/connectionmanager.go (hunk 1 above)
Comment (krnowak) [reply exporters/otlp/connectionmanager.go:50]: Hm, the reason I did it that way is that the hypothetical connection manager implementing the HTTP/JSON protocol could probably do it without pulling a dependency on protobuf. Not sure if this is worth the effort - it would mean writing a transformation from SDK data into JSON. What do you think?

Comment (krnowak) [reply exporters/otlp/connectionmanager.go:50]: One more note: if we decide to move the transformation back to exporter, then the name "ConnectionManager" would convey its purpose pretty accurately. On the other hand, if we keep it where it is now, then maybe "ProtocolDriver" name would be better.

Comment (MrAlias) [reply exporters/otlp/connectionmanager.go:50]: > the reason I did it that way is that the hypothetical connection manager implementing the HTTP/JSON protocol could probably do it without pulling a dependency on protobuf. Not sure if this is worth the effort - it would mean writing a transformation from SDK data into JSON.

That makes sense. Agreed, this sounds like a preferred approach.

> if we keep it where it is now, then maybe "ProtocolDriver" name would be better.

I like this. +1 to keeping the transform here and calling this ProtocolDriver. :+1:

File: exporters/otlp/grpcsingle.go [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+		m.metricsClient = colmetricpb.NewMetricsServiceClient(cc)
+		m.tracesClient = coltracepb.NewTraceServiceClient(cc)
+	} else {
+		m.metricsClient = nil
+		m.tracesClient = nil
+	}
+	return nil
+}
+
+func NewGRPCSingleConnectionManager(cfg GRPCConnectionConfig) ConnectionManager {
```
Comment (dashpole) [inline exporters/otlp/grpcsingle.go:70]: I think it is slightly easier to use if this takes ...options instead of a configuration since it doesn't require defining the configuration in a separate statement.

I.E.
```golang
func NewGRPCSingleConnectionManager(opts ...GRPCConnectionOption) ConnectionManager
```

Comment (dashpole) [inline exporters/otlp/grpcsingle.go:70]: `NewGRPCSingleConnectionManager` implies there is something other than "Single" that we would want to add in the future.  Would we want to make the "other" thing we want to add an option instead of a new function, and shorten this to `NewGRPCConnectionManager`?  That would be fine so long as the options would still apply, which I think they should.

**Comment by krnowak** [comment]
@dashpole: Thanks for the review. I'll reply to both your issues here.

> `NewGRPCSingleConnectionManager` implies there is something other than "Single" that we would want to add in the future. Would we want to make the "other" thing we want to add an option instead of a new function, and shorten this to `NewGRPCConnectionManager`? That would be fine so long as the options would still apply, which I think they should.

> I think it is slightly easier to use if this takes ...options instead of a configuration since it doesn't require defining the configuration in a separate statement.
> 
> I.E.
> 
> ```go
> func NewGRPCSingleConnectionManager(opts ...GRPCConnectionOption) ConnectionManager
> ```

The follow-up PR would introduce the `NewGRPCMultiConnectionManager` function, hence the `Single` in the name. The new function would create a connection manager that maintains separate connections for tracing and metrics. Both connections could be configured separately, which leads to a problem with the `opts ...GRPCConnectionOption` part: I don't see an elegant solution to configuring traces and metrics connection separately with the `opts ...GRPCConnectionOption` interface, other than duplicating all the options into `WithTracesFoo` and `WithMetricsFoo`. So for `NewGRPCMultiConnectionManager` I went with:

```go
type MultiConnectionConfig struct {
	Traces GRPCConnectionConfig
	Metrics GRPCConnectionConfig
}

func NewGRPCMultiConnectionManager(cfg MultiConnectionConfig) ConnectionManager { … }
```

So for `NewGRPCSingleConnectionManager` I went with taking `GRPCConnectionConfig` as a parameter, instead of `opts ...GRPCConnectionConfig` for consistency. This is rather inconsistent with the rest of the project, which is something that #536 seeks to resolve.

**Comment by dashpole** [comment]
> This is rather inconsistent with the rest of the project, which is something that #536 seeks to resolve.

Got it.  Thanks for linking the issue.

> The follow-up PR would introduce the NewGRPCMultiConnectionManager function

That makes sense.  This is part of https://github.com/open-telemetry/opentelemetry-go/issues/1121, right?

If there are eventually going to be multiple types of exporters (not just grpc), it might make sense to make a more generic single vs multi connection manager that isn't grpc-specific.  For example (with the current naming):

```golang
type MultiConnectionConfig struct {
	Traces ConnectionManager
	Metrics ConnectionManager
}

func NewMultiConnectionManager(cfg MultiConnectionConfig) ConnectionManager { … }
```

**Comment by krnowak** [comment]
> > This is rather inconsistent with the rest of the project, which is something that #536 seeks to resolve.
> 
> Got it. Thanks for linking the issue.
> 
> > The follow-up PR would introduce the NewGRPCMultiConnectionManager function
> 
> That makes sense. This is part of #1121, right?

Yes.

> 
> If there are eventually going to be multiple types of exporters (not just grpc), it might make sense to make a more generic single vs multi connection manager that isn't grpc-specific. For example (with the current naming):
> 
> ```go
> type MultiConnectionConfig struct {
> 	Traces ConnectionManager
> 	Metrics ConnectionManager
> }
> 
> func NewMultiConnectionManager(cfg MultiConnectionConfig) ConnectionManager { … }
> ```

I got something like that already: https://github.com/krnowak/opentelemetry-go/commit/f8c2ec8a3f48067c565956623ab44d8637ff5ae6, but maybe your variant with the config is actually better, because you spell you explicitly which manager goes for tracing and which for metrics. So thanks for the idea and we'll see how it fares during the review when it appears in another PR.

**Comment by krnowak** [comment]
Updated.

File: exporters/otlp/protocoldriver.go [hunk 4]
```diff
@@ -0,0 +1,51 @@
...
+
+	metricsdk "go.opentelemetry.io/otel/sdk/export/metric"
+	tracesdk "go.opentelemetry.io/otel/sdk/export/trace"
+)
+
+// ProtocolDriver is an interface used by OTLP exporter. It's
+// responsible for connecting to and disconnecting from the collector,
+// and for transforming traces and metrics into wire format and
+// transmitting them to the collector.
+type ProtocolDriver interface {
```
Comment (jmacd) [inline exporters/otlp/protocoldriver.go:28]: Looks good!

**Review by jmacd — APPROVED** [review]

**Comment by krnowak** [comment]
Updated. Added some tests to increase the coverage a bit (increasing coverage for error conditions is rather annoying). And fixed one or two races.

**Comment by krnowak** [comment]
Before merging it, I'd like to discuss one thing at the Golang SIG meeting first.

**Review by seanschade — COMMENTED** [review]
Just added a recommendation for a `NewGRPCConnectionConfig` function.

**Comment by MrAlias** [comment]
From the SIG meeting, we are going to not add a single/multiple gRPC connection driver. Instead going to progress to just using the driver interface to support multiple protocols for now.

**Comment by jmacd** [comment]
This looks good to me, independent of the single- vs multi-driver question, simply for taking us one step closer to an OTLP/HTTP exporter. 😀

File: exporters/otlp/grpcoptions.go [hunk 5]
```diff
@@ -0,0 +1,158 @@
...
+}
+
+type GRPCConnectionOption func(cfg *GRPCConnectionConfig)
+
+func (cfg GRPCConnectionConfig) Apply(opts ...GRPCConnectionOption) GRPCConnectionConfig {
+	cfg.ApplyInPlace(opts...)
+	return cfg
+}
+
+func (cfg *GRPCConnectionConfig) ApplyInPlace(opts ...GRPCConnectionOption) {
```
[the remark this answers is not in the document]
Comment (krnowak) [reply exporters/otlp/grpcoptions.go]: I simplified the code, so now the options are passed directly to `NewGRPCDriver`. The config struct is not exported any more.

**Comment by krnowak** [comment]
Rebased and updated. Simplified the gRPC driver a bit.

**Comment by krnowak** [comment]
Updated to fix a conflict in changelog. Updated some tests to be a bit less flaky.

**Review by MrAlias — APPROVED** [review]
Awesome work :+1:
